libraries/chain/hotstuff/finalizer.cpp

```diff
@@ -0,0 +1,302 @@
+#include <eosio/chain/hotstuff/finalizer.hpp>
+#include <eosio/chain/exceptions.hpp>
+#include <fc/log/logger_config.hpp>
+
+namespace eosio::chain {
+
+// ----------------------------------------------------------------------------------------
+block_state_ptr get_block_by_height(const fork_database_if_t::branch_type& branch, uint32_t block_num) {
+   auto it = std::find_if(branch.begin(), branch.end(),
+                          [&](const block_state_ptr& bsp) { return bsp->block_num() == block_num; });
+   return it == branch.end() ? block_state_ptr{} : *it;
+}
+
+// ----------------------------------------------------------------------------------------
+qc_chain_t finalizer::get_qc_chain(const block_state_ptr& proposal, const branch_type& branch) const {
+   qc_chain_t res;
+
+   // get b2
+   // ------
+   auto it2 = std::find_if(branch.begin(), branch.end(),
+        [t=proposal->core.last_qc_block_num](const block_state_ptr& bsp) { return bsp->block_num() == t; });
+   if (it2 == branch.end())
+      return res;
+   res.b2 = *it2;
+
+   // get b1
+   // ------
+   auto it1 = std::find_if(++it2, branch.end(),
+        [t=res.b2->core.last_qc_block_num](const block_state_ptr& bsp) { return bsp->block_num() == t; });
+   if (it1 == branch.end())
+      return res;
+   res.b1 = *it1;
+
+   // get b
+   // ------
+   auto it = std::find_if(++it1, branch.end(),
+        [t=res.b1->core.last_qc_block_num](const block_state_ptr& bsp) { return bsp->block_num() == t; });
+   if (it == branch.end())
+      return res;
+   res.b = *it;
+
+   return res;
+}
+
+// ----------------------------------------------------------------------------------------
+bool extends(const fork_database_if_t& fork_db, const block_state_ptr& descendant, const block_id_type& ancestor) {
+   if (ancestor.empty())
+      return false;
+   auto cur = fork_db.get_block_header(descendant->previous()); // use `get_block_header` so we can get the root
+   while (cur) {
+      if (cur->id() == ancestor)
+         return true;
+      cur = fork_db.get_block_header(cur->previous());
+   }
+   return false;
+}
+
+// ----------------------------------------------------------------------------------------
+finalizer::VoteDecision finalizer::decide_vote(const block_state_ptr& p, const fork_database_if_t& fork_db) {
```

Comment:
maybe use `bsp` for `p` as parameter name for clearer understanding.

Reply:
renamed it proposal.